File: api/src/services/audit.service.ts
/**
 * Audit service. Records every user action, agent action, and approval
 * decision. Append-only from the caller's perspective.
 */
import { v4 as uuid } from 'uuid';
import { getRepository } from '../persistence/factory';
import { AuditLog } from '../models';

const repo = () => getRepository<AuditLog>('auditLogs');

export interface AuditInput {
  actorType: AuditLog['actorType'];
  actor: string;
  action: string;
  targetType: string;
  targetId: string;
  correlationId?: string;
  workflowRunId?: string;
  approvalGateId?: string;
  details?: Record<string, unknown>;
}

export async function record(input: AuditInput): Promise<AuditLog> {
  const now = new Date();
  const entry: AuditLog = {
    id: uuid(),
    day: now.toISOString().slice(0, 10),
    timestamp: now.toISOString(),
    ...input,
  };
  return repo().upsert(entry);
}

export async function list(limit = 200): Promise<AuditLog[]> {
  const all = await repo().getAll();
  return all
    .sort((a, b) => b.timestamp.localeCompare(a.timestamp))
    .slice(0, limit);
}

export async function listForWorkflowRun(workflowRunId: string): Promise<AuditLog[]> {
  const rows = await repo().find((r) => r.workflowRunId === workflowRunId);
  return rows.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
}

File: api/app/audit_service.py
"""Audit service — append-only event log (mirror of audit.service.ts)."""
from typing import Dict, List, Optional

from .persistence import get_repository
from .util import new_id, now_iso


def _repo():
    return get_repository("auditLogs")


def record(
    *,
    actor_type: str,
    actor: str,
    action: str,
    target_type: str,
    target_id: str,
    correlation_id: Optional[str] = None,
    workflow_run_id: Optional[str] = None,
    approval_gate_id: Optional[str] = None,
    details: Optional[Dict] = None,
) -> Dict:
    now = now_iso()
    entry = {
        "id": new_id(),
        "day": now[:10],
        "timestamp": now,
        "actorType": actor_type,
        "actor": actor,
        "action": action,
        "targetType": target_type,
        "targetId": target_id,
        "correlationId": correlation_id,
        "workflowRunId": workflow_run_id,
        "approvalGateId": approval_gate_id,
        "details": details,
    }
    return _repo().upsert(entry)


def list_recent(limit: int = 200) -> List[Dict]:
    rows = sorted(_repo().get_all(), key=lambda r: r.get("timestamp", ""), reverse=True)
    return rows[:limit]


def list_for_workflow_run(workflow_run_id: str) -> List[Dict]:
    rows = _repo().find(lambda r: r.get("workflowRunId") == workflow_run_id)
    return sorted(rows, key=lambda r: r.get("timestamp", ""), reverse=True)
